apps/astarte_realm_management/lib/astarte_realm_management/engine.ex

```
@@ -273,43 +273,64 @@ defmodule Astarte.RealmManagement.Engine do
 
-            _ ->
-              {:halt, {:error, :incompatible_endpoint_change}}
-          end
-        end)
+      with {:ok, updated_mappings} <- extract_changed_mappings(mappings_map, existing_mappings) do
+        {:ok, Map.merge(new_mappings, updated_mappings)}
+      end
+    end
+  end
 
-      if is_map(maybe_new_mappings) do
-        {:ok, maybe_new_mappings}
+  defp extract_changed_mappings(old_mappings, existing_mappings) do
+    Enum.reduce_while(old_mappings, {:ok, %{}}, fn {mapping_id, old_mapping}, {:ok, acc} ->
+      with {:ok, updated_mapping} <- Map.fetch(existing_mappings, mapping_id),
+           {:allowed, true} <- {:allowed, allowed_mapping_update?(updated_mapping, old_mapping)},
+           {:updated, true} <- {:updated, is_mapping_updated?(updated_mapping, old_mapping)} do
+        {:cont, {:ok, Map.put(acc, mapping_id, updated_mapping)}}
       else
-        maybe_new_mappings
+        :error ->
+          {:halt, {:error, :missing_endpoints}}
+
+        {:allowed, false} ->
+          {:halt, {:error, :incompatible_endpoint_change}}
+
+        {:updated, false} ->
+          {:cont, {:ok, acc}}
       end
-    end
+    end)
+  end
+
+  defp allowed_mapping_update?(upd_mapping, mapping) do
+    new_mapping = drop_mapping_negligible_fields(upd_mapping)
+    old_mapping = drop_mapping_negligible_fields(mapping)
+
+    new_mapping == old_mapping
+  end
+
+  defp is_mapping_updated?(upd_mapping, mapping) do
+    mapping.explicit_timestamp != upd_mapping.explicit_timestamp or
+      mapping.doc != upd_mapping.doc or
+      mapping.description != upd_mapping.description
   end
 
-  defp drop_mapping_doc(%Mapping{} = mapping) do
-    %{mapping | description: nil, doc: nil}
+  defp drop_mapping_negligible_fields(%Mapping{} = mapping) do
+    %{mapping | doc: nil, description: nil, explicit_timestamp: false}
   end
 
-  defp drop_mapping_doc(nil) do
+  defp drop_mapping_negligible_fields(nil) do
     nil
   end
```

Is there a legitimate case where a mapping can be `nil`? Otherwise I'd remove this clause and let this crash if something strange happens